File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/theme/UseSmileIDSampleTheme.kt
package com.usesmileid.sampleapps.ui.theme

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Shapes
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.ui.unit.dp
import com.smileid.designsystem.SmileDimens

/** The Compose theme, built entirely from the vendored design-system tokens. */
@Composable
fun UseSmileIDSampleTheme(darkTheme: Boolean = isSystemInDarkTheme(), content: @Composable () -> Unit) {
    val colors = if (darkTheme) darkColors else lightColors
    CompositionLocalProvider(LocalUseSmileIDSampleColors provides colors) {
        MaterialTheme(
            colorScheme = colors.toMaterialColorScheme(darkTheme),
            shapes = shapes,
            typography = typography,
            content = content,
        )
    }
}

/** Token accessors for anything Material 3 has no slot for. */
object UseSmileIDSampleTheme {
    val colors: UseSmileIDSampleColors
        @Composable @ReadOnlyComposable get() = LocalUseSmileIDSampleColors.current

    val dimens = SmileDimens

    /** The full ramp, for the styles Material has no slot for. */
    val type = smileTypeStyles

    val shapes = UseSmileIDSampleShapes
}

/**
 * One shape per named surface. The radius comes from [SmileDimens]; this fixes where it is applied.
 * Everything that chooses a radius chooses 16 — `radius.sheet`'s 20 and `radius.xl`'s 20 are no
 * longer referenced by app code, though both survive in the generated token file.
 */
object UseSmileIDSampleShapes {
    val card = RoundedCornerShape(SmileDimens.radiusSurface)
    val tile = RoundedCornerShape(SmileDimens.radiusLg)

    /** The verifications row's tile: node 5206-2410 draws 10, the one radius no token carries. */
    val rowTile = RoundedCornerShape(ROW_TILE_RADIUS)
    val field = RoundedCornerShape(SmileDimens.radiusField)
    val pill = RoundedCornerShape(SmileDimens.radiusPill)
    val chip = RoundedCornerShape(SmileDimens.radiusChip)
    val sheet = RoundedCornerShape(topStart = SmileDimens.radiusSurface, topEnd = SmileDimens.radiusSurface)
}

/** PVR-A13 left verifications out of the migration to 16 deliberately. */
private val ROW_TILE_RADIUS = 10.dp

/** The SDK maps the same tokens onto the same slots, keeping host chrome and flow continuous. */
private fun UseSmileIDSampleColors.toMaterialColorScheme(darkTheme: Boolean) = if (darkTheme) {
    darkColorScheme(
        primary = primary,
        onPrimary = onPrimary,
        secondary = secondary,
        tertiary = accent,
        background = background,
        onBackground = textTitle,
        surface = surface,
        onSurface = textTitle,
        surfaceVariant = surfaceAlt,
        onSurfaceVariant = textBody,
        outline = border,
        error = errorFill,
        onError = onError,
        scrim = overlayScrim,
    )
} else {
    lightColorScheme(
        primary = primary,
        onPrimary = onPrimary,
        secondary = secondary,
        tertiary = accent,
        background = background,
        onBackground = textTitle,
        surface = surface,
        onSurface = textTitle,
        surfaceVariant = surfaceAlt,
        onSurfaceVariant = textBody,
        outline = border,
        error = errorFill,
        onError = onError,
        scrim = overlayScrim,
    )
}

private val shapes = Shapes(
    extraSmall = RoundedCornerShape(SmileDimens.radiusSm),
    small = RoundedCornerShape(SmileDimens.radiusSm),
    medium = RoundedCornerShape(SmileDimens.radiusMd),
    large = RoundedCornerShape(SmileDimens.radiusLg),
    extraLarge = RoundedCornerShape(SmileDimens.radiusSurface),
)

File: flutter/sample_ui/lib/src/theme/use_smileid_sample_theme.dart
import 'package:flutter/material.dart';

import '../tokens/smile_tokens.dart';
import 'use_smileid_sample_colors.dart';
import 'use_smileid_sample_typography.dart';

/// One shape per named surface; the radius comes from the tokens and this fixes where it is applied.
abstract final class UseSmileIDSampleShapes {
  /// A card, and the form's section surfaces.
  static final BorderRadius card = BorderRadius.circular(SmileDimens.radiusSurface);

  /// A square tile.
  static final BorderRadius tile = BorderRadius.circular(SmileDimens.radiusLg);

  /// The verifications row's tile: node 5206-2410 draws 10, the one radius no token carries.
  static const BorderRadius rowTile = BorderRadius.all(Radius.circular(10));

  /// A text field or a search field.
  static final BorderRadius field = BorderRadius.circular(SmileDimens.radiusField);

  /// A pill: the primary button, the status badges and the nav bar.
  static final BorderRadius pill = BorderRadius.circular(SmileDimens.radiusControl);

  /// A chip.
  static final BorderRadius chip = BorderRadius.circular(SmileDimens.radiusChip);

  /// A bottom sheet, rounded at the top only.
  static final BorderRadius sheet = BorderRadius.vertical(
    top: Radius.circular(SmileDimens.radiusSurface),
  );

  /// The avatar's rounded square — a square at radius 12, not a circle (node 5206:2904).
  static const BorderRadius avatar = BorderRadius.all(Radius.circular(12));
}

/// The app's theme, built entirely from the vendored design-system tokens.
abstract final class UseSmileIDSampleTheme {
  /// The light theme.
  static ThemeData light() => _themeFor(Brightness.light, UseSmileIDSampleColorSchemes.light);

  /// The dark theme.
  static ThemeData dark() => _themeFor(Brightness.dark, UseSmileIDSampleColorSchemes.dark);

  /// The component tokens for the active mode; every widget reads colours through this.
  static UseSmileIDSampleColors colorsOf(BuildContext context) =>
      Theme.of(context).extension<UseSmileIDSampleColors>() ??
      UseSmileIDSampleColorSchemes.light;
}

/// The SDK maps the same tokens onto the same roles, keeping host chrome and flow continuous.
ThemeData _themeFor(Brightness brightness, UseSmileIDSampleColors colors) {
  final ColorScheme scheme = ColorScheme(
    brightness: brightness,
    primary: colors.primary,
    onPrimary: colors.onPrimary,
    secondary: colors.secondary,
    onSecondary: colors.textInverse,
    tertiary: colors.accent,
    onTertiary: colors.textInverse,
    surface: colors.surface,
    onSurface: colors.textTitle,
    surfaceContainerHighest: colors.surfaceAlt,
    onSurfaceVariant: colors.textBody,
    outline: colors.border,
    error: colors.errorFill,
    onError: colors.onError,
    scrim: colors.overlayScrim,
  );
  return ThemeData(
    useMaterial3: true,
    brightness: brightness,
    colorScheme: scheme,
    scaffoldBackgroundColor: colors.background,
    fontFamily: useSmileIDSampleFontFamily,
    textTheme: _textTheme(colors),
    extensions: <ThemeExtension<dynamic>>[colors],
  );
}

/// The ramp on Material's own slots, reusing the nearest style where Material has none.
TextTheme _textTheme(UseSmileIDSampleColors colors) => TextTheme(
  displayLarge: UseSmileIDSampleType.textStyleDisplayLg,
  displayMedium: UseSmileIDSampleType.textStyleDisplayMd,
  displaySmall: UseSmileIDSampleType.textStyleHeadingPage,
  headlineLarge: UseSmileIDSampleType.textStyleHeadingPage,
  headlineMedium: UseSmileIDSampleType.textStyleHeadingCard,
  headlineSmall: UseSmileIDSampleType.textStyleHeadingSection,
  titleLarge: UseSmileIDSampleType.textStyleTitle,
  titleMedium: UseSmileIDSampleType.textStyleSubtitle,
  titleSmall: UseSmileIDSampleType.textStyleSubtitle,
  bodyLarge: UseSmileIDSampleType.textStyleBody,
  bodyMedium: UseSmileIDSampleType.textStyleBodySm,
  bodySmall: UseSmileIDSampleType.textStyleCaption,
  labelLarge: UseSmileIDSampleType.textStyleButton,
  labelMedium: UseSmileIDSampleType.textStyleButtonSm,
  labelSmall: UseSmileIDSampleType.textStyleOverline,
).apply(bodyColor: colors.textBody, displayColor: colors.textTitle);
